src/libcollections/binary_heap.rs

```diff
@@ -1189,3 +1218,49 @@ impl<'a, T: 'a + Ord + Copy> Extend<&'a T> for BinaryHeap<T> {
         self.extend(iter.into_iter().cloned());
     }
 }
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+pub struct PlaceIn<'a, T: 'a>
+where T: Clone + Ord {
+    heap: &'a mut BinaryHeap<T>,
+}
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+impl<'a, T> Place<T> for PlaceIn<'a, T>
+where T: Clone + Ord {
+    fn pointer(&mut self) -> *mut T {
+        self.heap.data.place_back().pointer()
+    }
+}
+
+#[unstable(feature = "collection_placement",
+           reason = "placement protocol is subject to change",
+           issue = "30172")]
+impl<'a, T> Placer<T> for PlaceIn<'a, T>
+where T: Clone + Ord {
+    type Place = PlaceIn<'a, T>;
+
+    fn make_place(self) -> Self {
+        let _ = self.heap.data.place_back().make_place();
+        self
```

Comment:
Implementation of this is going in the right direction but not quite there yet.

Probably the most notable issue here is that the `Place` that you create in the underlying vector gets immediatelly dropped, after which you are not allowed to put anything into the place in question anymore.

While this may work in current implementation, because `PlaceBack` for `Vec` has no `Drop` defined, consider what would happen if it *had* one:

```rust
// self.heap.data here is just big enough that a call to make_place would resize the vector.
let place = self.heap.data.place_back().make_place(); // resizes vector
drop(place); // resizes the vector back to the original size if was implemented weirdly
// now the placement protocol asks for a pointer to put data in
self.heap.data.place_back().pointer() // returns a pointer out of bounds!
```

This implementation therefore does not follow the placement protocol.

What ought be done instead is a new structure such as this (I think, feel free to explore alternative designs! generics omitted for brevity):

```rust
struct BinaryHeapPlace {
    heap: &mut BinaryHeap
    place: vec::PlaceBack
}
```

and then use the `place` in this structure to produce the `pointer` and also call the `finalize` on.